CONTRIBUTING.md

```diff
@@ -0,0 +1,187 @@
+
+## Preparing Your Commits
+
+Small PRs are easier to review and merge into the main branch. Minimize the scope of the changes in a single PR within reason. Changes that touch one or two files are likely to be reviewed and accepted more quickly than changes that touch every file in Flex.
+
+Format your commit messages following [Conventional Commits 1.0.0](https://www.conventionalcommits.org/en/v1.0.0/) (CC BY 3.0 netlify.com). Briefly:
+
+    <type>[(optional scope)][!]: <description>
+	[BLANK LINE]
+	[optional body]
+	[BLANK LINE]
+	[optional footer(s)]
+
+A typical commit message might look like:
+
+    docs: Add CONTRIBUTING manual
+	
+	Refs: #1234
+
+The description should summarize the changes in a single, short sentence. The description should be imperative, like the descriptions of the types below.
+
+The primary types are:
+- build: Change the build system, including tracking external dependencies
+- ci: Change the CI/CD system
+- docs: Update the package documentation
+- feat: Add a new feature
+- fix: Fix a bug or incorrect behavior
+- perf: Improve performance of a feature
+- refactor: Simplify maintainability without affecting performance or behavior
+- revert: Revert a previous commit
+- style: Correct a style mismatch (indentation, etc.)
+- test: Change or add to the test suite
+
+The optional scope must appear in parentheses and must consist of a noun describing the section of the code base that was changed.
+
+    docs(contrib): Update URLs in CONTRIBUTING manual
+
+The optional `!` must appear before the description whenever the commit introduces a breaking change - one that breaks compatibility with previous versions of the code base. Whenever the `!` marker appears, a BREAKING-CHANGE footer should also be included.
+
+The optional body may be included to provided additional information or context about the change. If it appears, it must be preceded by a blank line.
+
+The optional footers may be included to provide additional referential information and links to tracked issues, PRs, etc. Each footer must be preceded by a blank line. Footers must be formatted as `<footer-type>: <description>`, similar to the first line of commit messages. The description's format depends upon the footer-type. Known footer-types include:
+- Refs: A comma-separated list of commit hashes and/or issue and PR numbers. Issue and PR numbers must be prefixed with a `#`.
+- BREAKING-CHANGE: A description of the change that breaks compatibility with previous versions of the code, including the version number at which compatibility is broken. 
+
+A breaking change commit message might look like:
+
+    feat!: Switch to quantum scanner
+    
+    BREAKING-CHANGE: New scanner runs in constant time but doesn't support any existing hardware. 
+	Breaks compatibility with 1.0.0.
+
+Do not squash your commits unless requested by the maintainer. This aids discussion and revision during the review process.
```

Comment:
Hmm. Not quite, here. I'd say the default is to squash down to atomic commits. That is, if commits can go together, they should, but if separating commits makes it easier to separate out things for review, discussion or logic, then do that. In particular, what I don't want is a bunch of small commits that do nothing more than fix typos, adjust buffer size by 42.47, or add that last conditional that needed thought over night, even though it really went with the previous 105 things that got changed in the one file.

Reply:
I think developers should apply some common sense regarding whether they squash commits. There are cases where the commits should be squashed. E.g. when a commit is a "fixup" of a logic error or typo of a previous commit _of the same PR_, then it would be easier for others to review the code if the commits are squashed to one (so that we won't waste time reviewing a buggy revision that that developer has fixed already).